h2o-algos/src/main/java/hex/glm/GLM.java

```diff
@@ -1293,88 +1273,108 @@ else if(_parms._lambda_search) { // lambda search prefers coordinate descent
-        int k = 0;
-        for (int j = 0; j < x.length; ++j) {
-          if (i != j && x[j] != 0) {
-            x[k] = x[j];
-            nzs_ary[k++] = j;
+      for (int i = 0; i < activeData._cats; ++i) {
+        int var_min = activeData._catOffsets[i];
+        int var_max = activeData._catOffsets[i + 1];
+        for(int l = var_min; l < var_max; ++l) {
+          int k = 0;
+          double [] x = xx[l];
+          for (int j = 0; j < var_min; ++j)
+            if (x[j] != 0) nzs_ary[k++] = j;
+          for (int j = var_max; j < activeData.numStart(); ++j)
+            if (x[j] != 0) nzs_ary[k++] = j;
+          if (k < ((nzs_ary.length - var_max + var_min) >> 3)) {
+            sparseCnt++;
+            nzs[l] = Arrays.copyOf(nzs_ary, k);
           }
         }
-        if (k < (nzs_ary.length >> 3)) {
-          nzs[i] = Arrays.copyOf(nzs_ary, k);
-          xx[i] = Arrays.copyOf(x,k);
-        }
       }
     }
-    double [] grads = new double [gt._xy.length];
+    Log.info("COD::nzs done in " + (System.currentTimeMillis()-t0) + "ms, found " + sparseCnt + " sparse columns");
+    final BetaConstraint bc = _state.activeBC();
     double [] beta = _state.beta().clone();
-    for(int i = 0; i < grads.length; ++i) {
-      double ip = 0;
-      if(i < nzs.length && nzs[i] != null) {
-        int [] ids = nzs[i];
-        double [] x = xx[i];
-        for(int j = 0; j < nzs[i].length; ++j)
-          ip += x[j]*beta[ids[j]];
-        grads[i] =  gt._xy[i] - ip;
-      } else {
-        grads[i] =  gt._xy[i] - ArrayUtils.innerProduct(xx[i], beta) + xx[i][i] * beta[i];
+    int numStart = activeData.numStart();
+    if(newCols != null) {
+      for (int id : newCols) {
+        double b = bc.applyBounds(ADMM.shrinkage(grads[id], l1pen) * diagInv[id], id);
+        if (b != 0) {
+          doUpdateCD(grads, xx[id], -b, id, id + 1);
+          beta[id] = b;
+        }
       }
     }
     int iter1 = 0;
-    int P = gt._xy.length - 1;
-    final BetaConstraint bc = _state.activeBC();
-    DataInfo activeData = _state.activeData();
+    int P = xy.length - 1;
     // CD loop
-    while (iter1++ < 1000 /*Math.max(P,500)*/) {
-      double bdiffPos = 0;
-      double bdiffNeg = 0;
+    long t2 = System.currentTimeMillis();
+//    // CD loop
+    while (iter1++ < Math.max(P,500)) {
+      double maxDiff = 0;
       for (int i = 0; i < activeData._cats; ++i) {
         for(int j = activeData._catOffsets[i]; j < activeData._catOffsets[i+1]; ++j) { // can do in parallel
           double b = bc.applyBounds(ADMM.shrinkage(grads[j], l1pen) * diagInv[j],j);
           double bd = beta[j] - b;
-          bdiffPos = bd > bdiffPos?bd:bdiffPos;
-          bdiffNeg = bd < bdiffNeg?bd:bdiffNeg;
-          if(nzs[j] == null)
-            doUpdateCD(grads, xx[j], bd, activeData._catOffsets[i], activeData._catOffsets[i + 1]);
-          else
-            doSparseUpdateCD(grads, xx[j], nzs[j], bd, activeData._catOffsets[i], activeData._catOffsets[i + 1]);
```

Comment:
I think the doSparseUpdateCD() method should be deleted, it doesn't seem to be used anymore